examples/stream/client.c

```diff
@@ -0,0 +1,263 @@
+    assert(job != NULL);
+
+    /* Setup buffers */
+    rs_buffers_t bufs = { 0 };
+    bufs.next_in = in_buf;
+    bufs.next_out = out_buf;
+    bufs.avail_out = sizeof(out_buf);
+
+    /* Generate signature */
+    do {
+        if (bufs.eof_in == 0) {
+            if (bufs.avail_in > 0) {
+                /* Leftover tail data, move it to front */
+                memmove(in_buf, bufs.next_in, bufs.avail_in);
+            }
+
+            /* Fill input buffer */
+            size_t n_bytes = fread(in_buf + bufs.avail_in, 1, sizeof(in_buf) - bufs.avail_in, file);
+            if (n_bytes == 0) {
+                if (ferror(file)) {
+                    perror("Failed to read file");
+                    rs_file_close(file);
+                    rs_job_free(job);
+                    return -1;
+                }
+
+                /* End-of-File reached */
+                bufs.eof_in = feof(file);
+                assert(bufs.eof_in != 0);
+            }
+
+            bufs.next_in = in_buf;
+            bufs.avail_in += n_bytes;
+        }
+
+        /* Iterate job */
+        res = rs_job_iter(job, &bufs);
+        if (res != RS_DONE && res != RS_BLOCKED) {
+            rs_file_close(file);
+            rs_job_free(job);
+            return -1;
+        }
+
+        size_t present = bufs.next_out - out_buf;
+        if (present > 0) {
+            /* Drain output buffer */
+            int ret = send_message(sock, out_buf, present, (res == RS_DONE) ? 1 : 0);
+            if (ret == -1) {
+                rs_file_close(file);
+                rs_job_free(job);
+                return -1;
+            }
+
+            bufs.next_out = out_buf;
+            bufs.avail_out = sizeof(out_buf);
+        }
+    } while (res != RS_DONE);
+
+    rs_file_close(file);
+    rs_job_free(job);
+
+    return 0;
+}
+
+static int recv_delta_and_patch_file(int sock, const char *fname) {
+    char fname_new[PATH_MAX];
+    int ret = snprintf(fname_new, sizeof(fname_new), "%s.new", fname);
+    if (ret < 0 || (size_t)ret >= sizeof(fname_new)) {
+        fputs("Filename too long\n", stderr);
+        return -1;
+    }
+
+    /* Open new file */
+    FILE *new = rs_file_open(fname_new, "wb", 1);
+    assert(new != NULL);
+
+    /* Open basis file */
+    FILE *old = old = rs_file_open(fname, "rb", 0);
+    assert(old != NULL);
+
+    rs_job_t *job = rs_patch_begin(rs_file_copy_cb, old);
+    assert(job != NULL);
+
+    /* Setup RSYNC buffers */
+    rs_buffers_t bufs = { 0 };
+    bufs.next_in = in_buf;
+    bufs.next_out = out_buf;
+    bufs.avail_out = sizeof(out_buf);
+
+    rs_result res;
+    do {
+        if (bufs.eof_in == 0) {
+            if (bufs.avail_in > 0) {
+                /* Left over tail data, move to front */
+                memmove(in_buf, bufs.next_in, bufs.avail_in);
+            }
+
+            size_t n_bytes;
+            int ret = recv_message(sock, in_buf + bufs.avail_in, &n_bytes, &bufs.eof_in);
```

Comment:
I just realized that if there is leftover tail data in the input buffer, the next call to `recv_message()` may write out of bounds of the array. I will fix this ;)